### PR TITLE
nephio-ansible-install

### PR DESCRIPTION
**Pull request by henderiw**
this adds a PR to install the Nephio environment for the one summit demo through a set of sensible playbooks

### CODE REVIEW

**Comment by johnbelamaric** [comment]
Nice! I'd like the other folks involved in the workshop setup to take a look and provide feedback. A couple high level things:

 - Does this replace the automation we used for the VM setup in the workshop?
 - It looks like the `participant` package installation is not part of these?
 - I think the copyright notice should read "The Nephio Authors" instead of "Nokia", similar to https://github.com/kubernetes/community/blob/master/github-management/kubernetes-repositories.md#sig-repositories - I imagine this relates to the CLA.

File: nephio-ansible-install/README.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+- 32G RAM, 8 vcpu -> we can change this based on the amount of kind clusters we need
+- SSH access with a SSH key is setup + username
+
+The creation of the VM is right now out of scope, but we can see what we can do going forward.
+Also we assume right now the ansible playbook is executed remote from the VM. We can see if people want to use a different approach going forward.
+
+In a local environment clone the repo in a local environment
+
+```bash
+git clone https://github.com/henderiw-nephio/nephio-ansible-install
```
Comment (vjayaramrh) [inline nephio-ansible-install/README.md]: This URL needs to be replaced with `https://github.com/nephio-project/one-summit-22-workshop.git`

File: nephio-ansible-install/README.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+- cloud_user: the username that is created to access the VM using SSH
+- github_username: your github user name
+- github_token: github access token to access github [github personal access token](https://docs.github.com/en/enterprise-server@3.4/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
+- github_organization: (optional) if you use a github organization for the repo's you should add your github organization here, otherwise it uses the github username
+- gitea_username: your [gitea](https://gitea.io) (local repo) username
+- gitea_password: your [gitea](https://gitea.io) (local repo) password
+
+Note: You can choose between using remote github repos or local gitea repos for your Nephio environment by setting either the github variables or the gitea variables.
+
+To start running ansible playbooks an ansible environment is required. Below is an example how to install ansible using a virtual environment. The repo scripts rely on the ansible galaxy community collection
```
Comment (vjayaramrh) [inline nephio-ansible-install/README.md]: Alternatively, ansible playbooks can be mounted inside of an ansible container and executed from there 

e.g; `sudo podman run --name ansible --net=host --cap-add=NET_ADMIN -it -v <path to ansible playbook folder>:/ansible/ -v ~/.kube:/root/.kube -v ~/.ssh:/root/.ssh willhallonline/ansible:2.11.10-centos-7 /bin/sh`

File: nephio-ansible-install/roles/tree/tasks/main.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+## © 2022 Nokia
+## Licensed under the Apache License 2.0
+## SPDX-License-Identifier: Apache-2.0
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/tree/tasks/main.yaml:3]: Do the above three lines need to be removed?

File: nephio-ansible-install/roles/topo/destroy/tasks/main.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+## © 2022 Nokia
+## Licensed under the Apache License 2.0
+## SPDX-License-Identifier: Apache-2.0
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/topo/destroy/tasks/main.yaml:3]: Do the above three lines need to be removed?

File: nephio-ansible-install/playbooks/install-prereq.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+- hosts: vm
+  environment:
+    http_proxy: "{{proxy.http_proxy}}"
+    https_proxy: "{{proxy.https_proxy}}"
+    no_proxy: "{{proxy.no_proxy}}"
+  roles:
+    
+    - { role: docker, tags: docker}
```
Comment (vjayaramrh) [inline nephio-ansible-install/playbooks/install-prereq.yaml:9]: having podman installed would be good as well, but can be part of a separate PR.
I shall open an issue to track installing podman

File: nephio-ansible-install/roles/kubectl/tasks/main.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    mode: 0600
+  tags:
+    - kubectl
+
+- name: Copy kubectl binary to destination directory
+  become: true
+  ansible.builtin.copy:
+    src: "{{ tmp_directory }}/{{ item }}"
+    dest: "{{ bin_directory }}/{{ item }}"
+    mode: "0755"
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/kubectl/tasks/main.yaml:20]: Nit: It could be useful to keep mode, owner and group to be parameterized as well for the long term, but OK to keep it hard-coded in the short term.

File: nephio-ansible-install/roles/cni/tasks/main.yaml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+- name: Create tmp cni directory
+  ansible.builtin.file:
+    path: "{{ tmp_directory }}/cni"
+    state: directory
+    mode: '0755'
+
+- name: Download cni archive
+  ansible.builtin.get_url:
+    url: https://github.com/containernetworking/plugins/releases/download/v{{ cni_version }}/cni-plugins-{{ host_os }}-{{ host_arch }}-v{{ cni_version }}.tgz
+    #url: "https://kind.sigs.k8s.io/dl/v{{ kind_version }}/kind-{{ host_os }}-{{ host_arch }}"
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/cni/tasks/main.yaml]: Commented lines L15 and L16 can be removed

File: nephio-ansible-install/roles/cni/tasks/main.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+- name: Unarchive cni archive
+  ansible.builtin.unarchive:
+    src: "{{ tmp_directory }}/cni-plugins-{{ host_os }}-{{ host_arch }}-v{{ cni_version }}.tgz"
+    dest: "{{ tmp_directory }}/cni"
+    remote_src: true
+  tags:
+    - cni
+
+#- name: Download cni binaries
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/cni/tasks/main.yaml]: WOuld be good to remove the below commented lines

File: nephio-ansible-install/roles/docker/tasks/main.yaml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  become: true
+  apt: name=docker-ce update_cache=yes
+
+- name: Add user to docker group
+  become: true
+  shell: usermod -aG docker {{ cloud_user }}
+
+- name: Reset connection so docker group is picked up.
+  meta: reset_connection
+- name: Create the docker network for kind (to make our job easier plumbing up the lot)
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/docker/tasks/main.yaml:27]: you will need a blank line after L25

File: nephio-ansible-install/roles/docker/tasks/main.yaml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  docker_network:
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/docker/tasks/main.yaml:28]: What pip3 module needs to be installed on the ansible controller to use this? I see the below error

```
vjayaram@wim-pr-test:~/nephio-ansible-install$ ansible-playbook -i 10.10.10.8, playbooks/install-prereq.yaml 
ERROR! couldn't resolve module/action 'docker_network'. This often indicates a misspelling, missing collection, or incorrect module path.

The error appears to be in '/home/vjayaram/nephio-ansible-install/roles/docker/tasks/main.yaml': line 27, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: Create the docker network for kind (to make our job easier plumbing up the lot)
  ^ here
```

File: nephio-ansible-install/README.md (hunk 1 above)
Comment (henderiw) [reply nephio-ansible-install/README.md]: fixed using: https://github.com/nephio-project/one-summit-22-workshop/pull/106/commits/a854bd5fcfaa5448a7d18b4d8742929f0fa059d6

File: nephio-ansible-install/README.md (hunk 2 above)
Comment (henderiw) [reply nephio-ansible-install/README.md]: yes this is one option we should provide and maybe add even a vagrant/packer file. Some people even suggested docker in docker for this.

File: nephio-ansible-install/playbooks/install-prereq.yaml (hunk 5 above)
Comment (henderiw) [reply nephio-ansible-install/playbooks/install-prereq.yaml:9]: yes it would be good to be able to select between both of them

File: nephio-ansible-install/roles/cni/tasks/main.yaml (hunk 8 above)
Comment (henderiw) [reply nephio-ansible-install/roles/cni/tasks/main.yaml]: removed comments using:
https://github.com/nephio-project/one-summit-22-workshop/pull/106/commits/48ea2f1523683674dad6749edd5f596da3109cc7

File: nephio-ansible-install/roles/docker/tasks/main.yaml (hunk 9 above)
Comment (henderiw) [reply nephio-ansible-install/roles/docker/tasks/main.yaml:27]: fixed: https://github.com/nephio-project/one-summit-22-workshop/pull/106/commits/395f7e39e539da55dfac934b347349d1f61f9f83

File: nephio-ansible-install/roles/nephio/install/tasks/packages_files.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+- name: Download the {{ item.package }} packages, with url {{ nephio.packages_url }}/{{ item.package }}
+  shell: kpt pkg get --for-deployment {{ nephio.packages_url }}/{{ item.package }}
+  args: 
+    chdir: "{{ tmp_directory }}/{{ nephio.install_dir}}"
+  register: result
+  failed_when:
+    - result.rc > 1
+    - result.rc == 1 and "already exists" not in result.stderr
+
+- name: Initialize {{ item.package }}
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/nephio/install/tasks/packages_files.yaml:15]: FYI I ran the playbook locally on the VM via the command

```
ansible-playbook playbooks/install-prereq.yaml --connection=local
```

and ran into the below issue for this taks

```
TASK [nephio/install : Initialize nephio-configsync] ********************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "kpt fn render nephio-configsync", "delta": "0:00:00.212946", "end": "2022-12-13 01:41:46.003805", "msg": "non-zero return code", "rc": 1, "start": "2022-12-13 01:41:45.790859", "stderr": "Package \"nephio-configsync\": \n[RUNNING] \"gcr.io/kpt-fn/apply-replacements:v0.1.1\"\n[FAIL] \"gcr.io/kpt-fn/apply-replacements:v0.1.1\" in 0s\nError: exit status 1\ndocker must be running to use this command\nTo install docker, follow the instructions at https://docs.docker.com/get-docker/.\n ", "stderr_lines": ["Package \"nephio-configsync\": ", "[RUNNING] \"gcr.io/kpt-fn/apply-replacements:v0.1.1\"", "[FAIL] \"gcr.io/kpt-fn/apply-replacements:v0.1.1\" in 0s", "Error: exit status 1", "docker must be running to use this command", "To install docker, follow the instructions at https://docs.docker.com/get-docker/.", " "], "stdout": "", "stdout_lines": []}
```

I had to log out of the VM and re-login to run the playbook again to go through all the roles/tasks in install_prereq.yaml playbook.

File: nephio-ansible-install/roles/gitea/create/tasks/main.yaml [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  when:
+    - gitea_username is defined
+    - gitea_password is defined
+- name: Create a gitea directory
+  ansible.builtin.file:
+    path: /home/{{ cloud_user}}/gitea
+    state: directory
+    mode: '0755'
+- name: Start gitea container
+  docker_container:
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/gitea/create/tasks/main.yaml:17]: seems like the module docker_container is not getting recognized when I run the playbook on Ubuntu VM (Ubuntu 22.04.1 LTS)

```
vjayaram@wim-pr-test3:~/nephio-ansible-install$ ansible-playbook playbooks/create-gitea.yaml --connection=local 
ERROR! couldn't resolve module/action 'community.general.docker_container'. This often indicates a misspelling, missing collection, or incorrect module path.

The error appears to be in '/home/vjayaram/nephio-ansible-install/roles/gitea/create/tasks/main.yaml': line 15, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: Start gitea container
  ^ here
```

I executed the  `ansible-galaxy collection install community.general` as well per README

File: nephio-ansible-install/README.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+To start running ansible playbooks an ansible environment is required. Below is an example how to install ansible using a virtual environment. The repository scripts rely on the ansible galaxy community collection
+
+```python
+python3 -m venv .venv
+source .venv/bin/activate
+pip install --upgrade pip
+pip install ansible
+pip install pygithub
+ansible-galaxy collection install community.general
```
Comment (vjayaramrh) [inline nephio-ansible-install/README.md:90]: We may also to call out that `ansible-galaxy collection install community.docker` needs to be executed as well. This would be needed wherever the docker_* module is being used

File: nephio-ansible-install/roles/gitea/create/tasks/main.yaml (hunk 12 above)
Comment (vjayaramrh) [reply nephio-ansible-install/roles/gitea/create/tasks/main.yaml:17]: I was able to resolve the issue when I ran `ansible-galaxy collection install community.docker`, so the README needs an update. I have left a comment in the README as well

File: nephio-ansible-install/roles/gitea/create/tasks/main.yaml [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+  when:
+    - gitea_username is defined
+    - gitea_password is defined
+- name: Wait for gitea to initialise
+  ansible.builtin.wait_for:
+    path: /home/{{ cloud_user }}/gitea/ssh
+  when:
+    - gitea_username is defined
+    - gitea_password is defined
+- name: Creating nephio user in gitea
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/gitea/create/tasks/main.yaml:41]: FYI Ran into the below when this task was getting executed

```
TASK [gitea/create : Creating nephio user in gitea] *********************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "docker exec -it gitea  gitea admin user create    --username nephio    --password nephio    --must-change-password=false    --access-token    --email=nephio@nephio.org > /home/vjayaram/gitea/nephio-create\n", "delta": "0:00:00.034046", "end": "2022-12-13 03:29:52.192736", "msg": "non-zero return code", "rc": 1, "start": "2022-12-13 03:29:52.158690", "stderr": "the input device is not a TTY", "stderr_lines": ["the input device is not a TTY"], "stdout": "", "stdout_lines": []}
```

File: nephio-ansible-install/roles/gitea/create/tasks/main.yaml [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+  shell: |
+    docker exec -it gitea \
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/gitea/create/tasks/main.yaml]: I was able to execute the playbook without the above error once I removed the `-it` ... did anyone else have issue running the playbook as is?

File: nephio-ansible-install/roles/repo-gitea/create/tasks/main.yaml [hunk 16]
```diff
@@ -0,0 +1,27 @@
+- name: Slurp access token
+  ansible.builtin.slurp:
+    src: /home/{{ cloud_user }}/gitea/nephio-gitea-token
+  register: gitea_token
+  when:
+    - gitea_username is defined
+- name: Create repos
```
Comment (vjayaramrh) [inline nephio-ansible-install/roles/repo-gitea/create/tasks/main.yaml:11]: I see the below when this task is being executed

```
failed: [localhost] (item={'key': 'mgmt', 'value': {'mgmt_subnet': '172.88.0.0/16', 'pod_subnet': '10.196.0.0/16', 'svc_subnet': '10.96.0.0/16'}}) => {"ansible_loop_var": "item", "cache_control": "no-store, no-transform", "changed": false, "connection": "close", "content_length": "74", "content_type": "application/json;charset=utf-8", "date": "Tue, 13 Dec 2022 16:12:02 GMT", "elapsed": 0, "item": {"key": "mgmt", "value": {"mgmt_subnet": "172.88.0.0/16", "pod_subnet": "10.196.0.0/16", "svc_subnet": "10.96.0.0/16"}}, "json": {"message": "token is required", "url": "http://localhost:3000/api/swagger"}, "msg": "Status code was 401 and not [201]: HTTP Error 401: Unauthorized", "redirected": false, "status": 401, "url": "http://localhost:3000/api/v1/user/repos", "x_content_type_options": "nosniff", "x_frame_options": "SAMEORIGIN"}
```

File: nephio-ansible-install/roles/gitea/create/tasks/main.yaml (hunk 15 above)
Comment (danaschwanden) [reply nephio-ansible-install/roles/gitea/create/tasks/main.yaml]: Hi @vjayaramrh , you are right. Removing the -t is doing the trick.
This would resolve the error message in your review comment above and also resolve the follow on error in your review comment below.

File: nephio-ansible-install/roles/repo-gitea/create/tasks/main.yaml (hunk 16 above)
Comment (danaschwanden) [reply nephio-ansible-install/roles/repo-gitea/create/tasks/main.yaml:11]: Hi @vjayaramrh , this is a follow on error from the issue you mentioned in your review comment above. Removing the -t in the docker exec will allow for that task to run which will unblock this one.

**Comment by henderiw** [comment]
@danaschwanden,  I believe there are some hardcoded usernames in the gatea deploy cluster-deploy role.
e.g. here. We have to change this to the gate_username variable afars, not?

- name: configure secret
  shell: kubectl --kubeconfig ~/.kube/mgmt-config create secret generic github-personal-access-token \
    --from-literal=username={{github_username}} \
    --from-literal=token={{github_token}}
  failed_when:
    - result.rc > 1
    - result.rc == 1 and "already exists" not in result.stderr
  when:
    - github_username is defined

File: nephio-ansible-install/roles/docker/tasks/main.yaml (hunk 9 above)
Comment (vjayaramrh) [inline nephio-ansible-install/roles/docker/tasks/main.yaml:27]: When this task is invoked as part of the install-prereq.yaml playbook, the below error is seen.
I was running the playbook locally using `ansible-playbook playbooks/install-prereq.yaml --connection=local`. I had to log out and log back in; then rerun the playbook again.
It might be a good idea to update the README with instructions that one has to a)log out and log back in b)re-run the playbook again

```
TASK [docker : Create the docker network for kind (to make our job easier plumbing up the lot)] ********************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Error connecting: Error while fetching server API version: ('Connection aborted.', PermissionError(13, 'Permission denied'))"}

PLAY RECAP *********************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=5    changed=4    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0  
```

File: nephio-ansible-install/roles/gitea/create/tasks/main.yaml (hunk 15 above)
Comment (vjayaramrh) [reply nephio-ansible-install/roles/gitea/create/tasks/main.yaml]: @danaschwanden In my case, I removed the entire `-it`

File: nephio-ansible-install/roles/repo-gitea/create/tasks/main.yaml (hunk 16 above)
Comment (vjayaramrh) [reply nephio-ansible-install/roles/repo-gitea/create/tasks/main.yaml:11]: @danaschwanden sounds good

**Comment by danaschwanden** [comment]
> @danaschwanden, I believe there are some hardcoded usernames in the gatea deploy cluster-deploy role. e.g. here. We have to change this to the gate_username variable afars, not?

Hi @henderiw , the proposed idea is to allow specifying the gitea username/passwd in the inventory file so gitea can be accessed through the GUI if needed. The gitea create task creates this user once the gitea container is up and running while also creating a gitea access token for the automated API access from the mgmt cluster (following the same approach as you do for github).

File: nephio-ansible-install/roles/repo-gitea/create/tasks/main.yaml (hunk 16 above)
Comment (vjayaramrh) [reply nephio-ansible-install/roles/repo-gitea/create/tasks/main.yaml:11]: When I ran again on a newly created VM along with @henderiw , I did not encounter this issue again. So, can ignore the above comment for now

File: nephio-ansible-install/roles/tree/tasks/main.yaml (hunk 3 above)
Comment (henderiw) [reply nephio-ansible-install/roles/tree/tasks/main.yaml:3]: I am discussing with some Nokia folks to remove this

File: nephio-ansible-install/playbooks/install-prereq.yaml (hunk 5 above)
Comment (henderiw) [reply nephio-ansible-install/playbooks/install-prereq.yaml:9]: lets do this in a separate PR

File: nephio-ansible-install/roles/nephio/install/tasks/packages_files.yaml (hunk 11 above)
Comment (henderiw) [reply nephio-ansible-install/roles/nephio/install/tasks/packages_files.yaml:15]: we need to add some text in the readme to use connection-local. This indeed requires you to login and logout again.

File: nephio-ansible-install/roles/gitea/create/tasks/main.yaml (hunk 12 above)
Comment (henderiw) [reply nephio-ansible-install/roles/gitea/create/tasks/main.yaml:17]: fixed https://github.com/nephio-project/one-summit-22-workshop/pull/106/commits/92affdf9ed2c2a920e7843266952f4bd89edc5f3

File: nephio-ansible-install/README.md (hunk 13 above)
Comment (henderiw) [reply nephio-ansible-install/README.md:90]: fixed https://github.com/nephio-project/one-summit-22-workshop/pull/106/commits/92affdf9ed2c2a920e7843266952f4bd89edc5f3

File: nephio-ansible-install/roles/gitea/create/tasks/main.yaml (hunk 14 above)
Comment (henderiw) [reply nephio-ansible-install/roles/gitea/create/tasks/main.yaml:41]: fixed: [5c9d912](https://github.com/nephio-project/one-summit-22-workshop/pull/106/commits/5c9d912f784f71a0de58bbea88ccc54bbbbb9cfe)

File: nephio-ansible-install/roles/gitea/create/tasks/main.yaml (hunk 15 above)
Comment (henderiw) [reply nephio-ansible-install/roles/gitea/create/tasks/main.yaml]: fixed [5c9d912](https://github.com/nephio-project/one-summit-22-workshop/pull/106/commits/5c9d912f784f71a0de58bbea88ccc54bbbbb9cfe)

File: nephio-ansible-install/README.md [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+```bash
+ansible-playbook playbooks/deploy-clusters.yaml
+```
+
+Lastly we install the environment manifests we use for the workshop scenario's
+
+```bash
+ansible-playbook playbooks/configure-nephio.yaml
+```
```
Comment (vjayaramrh) [inline nephio-ansible-install/README.md:132]: How about adding commands and expected output that can be executed to verify that installation has succeeded? We could also add screenshots of the nephio webui and/or gitea webui as well

File: nephio-ansible-install/roles/tree/tasks/main.yaml (hunk 3 above)
Comment (henderiw) [reply nephio-ansible-install/roles/tree/tasks/main.yaml:3]: fixed https://github.com/nephio-project/one-summit-22-workshop/pull/106/commits/30345990bf7948269fe4954e827cb8740fb3984d

File: nephio-ansible-install/roles/topo/destroy/tasks/main.yaml (hunk 4 above)
Comment (henderiw) [reply nephio-ansible-install/roles/topo/destroy/tasks/main.yaml:3]: fixed https://github.com/nephio-project/one-summit-22-workshop/pull/106/commits/30345990bf7948269fe4954e827cb8740fb3984d

File: nephio-ansible-install/roles/tree/tasks/main.yaml (hunk 3 above)
Comment (iwojdan) [reply nephio-ansible-install/roles/tree/tasks/main.yaml:3]: Hello 
I succeeded to install entire environment by those ansible scripts.
I started to perform exercise from one-summit-22 with UPF 5g core deployment.  But when I open web ui on local browser there is no button "add organizational blueprint" in screen Organizational Blueprints
